lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,231 @@
+require 'set'
+require 'yaml'
+require 'open-uri'
+require 'objspace'
+require 'fog/core'
+
+module Fog
+  module Orchestration
+    module Util
+      #
+      # Resolve get_file resources found in a HOT template populating
+      #  a files Hash conforming to Heat Specs
+      #  https://developer.openstack.org/api-ref/orchestration/v1/index.html?expanded=create-stack-detail#stacks
+      #
+      # This implementation just process nested templates but not resource
+      #  registries.
+      class RecursiveHotFileLoader
+        attr_reader :files
+        attr_reader :template
+        attr_reader :visited
+
+        def initialize(template, files = nil)
+          # Serialize the template hash to deep-copy it and
+          #  avoid modifying the argument. Instead create a
+          #  new one to be modified by get_file_contents.
+          @template = deep_copy(template)
+          @files = files || {}
+          @visited = Set.new
+        end
+
+        def get_files
```

Comment:
`get_files` method is not needed if we write `initialize` this way
```
def initialize(template)
  @files = {}
  @visited = Set.new
  @template = get_template_content(template)
end
```
In the caller you can simply
```
file_resolver = Fog::Orchestration::Util::RecursiveHotFileLoader.new(options[:template] || options[:template_url]
options[:template] = file_resolver.template
options[:files] = file_resolver.files unless file_resolver.files.empty?
```